KAFKA-3856: Refactoring for KIP-120

mjsax opened the pull request:
- extract InternalTopologyBuilder from TopologyBuilder
 - deprecate all "leaking" methods from public TopologyBuilder API
 - changed TopologyDescription and all nested classed into interfaces

mjsax committed:
KAFKA-3856: Refactoring for KIP-120
 - extract InternalTopologyBuilder from TopologyBuilder
 - deprecate all "leaking" methods from public TopologyBuilder API

mjsax commented:
Call for review @bbejeck @dguy @enothereska @guozhangwang

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6110/
Test PASSed (JDK 7 and Scala 2.11).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6095/
Test PASSed (JDK 8 and Scala 2.12).

enothereska commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/ProcessorTopologyTest.java:
```diff
@@ -222,9 +222,9 @@ public void shouldDriveGlobalStore() throws Exception {
 
     @Test
-    public void testDrivingSimpleMultiSourceTopology() {
+    public void testDrivingSimpleMultiSourceTopology() throws Exception {
```
Which part throws the exception? (same question for the other tests below). Could we use a specific exception, or is it broad?

mjsax replied:
I guess, this is some "left over" as I cut down this PR to make is smaller... (ie, did remove some other changes that will follow in another RP -- just forgot to undo this changes to -- should be ok to just leave it IMHO) For test in general it's ok to just declare "Exception" as nobody catches it anyway (it's best practice in Java AFAIK).

bbejeck replied:
+1 to leave in.  IMHO it makes the test more readable without the try/catch boilerplate.

enothereska commented:
Massive cleanup, thanks @mjsax. I couldn't find `org.apache.kafka.streams.TopologyDescription` as described in the KIP. Is this WiP?

mjsax commented:
@enothereska I'll split up the whole KIP in multiple PRs. Otherwise it's a mess to review. There is already #2301 that does add `TopologyDescription` -- but there will be at least 2 more PRs (one for adding `Topology` and one for adding `StreamsBuilder`). Small PRs FTW!

bbejeck commented:
Looks good, thanks for the big cleanup.  

One question (sorry if this is an obvious one): the `TopologyBuilder` has the `InternalTopologyBuilder` marked as deprecated. The deprecation indicates using the `InternalTopologyBuilder` from within `TopologyBuilder` is deprecated and not the `InternalTopologyBuilder` itself, correct?

guozhangwang commented:
@mjsax I'm wondering if could you try breaking this PR into to commits for easier reviewing? One commit for just extracting out the `InternalTopologyBuilder` from `TopologyBuilder` without the actual impl of the `InternalTopologyBuilder`, just to show what are the left public APIs of `TopologyBuilder` and what are the internal interfaces of `InternalTopologyBuilder`?

EDIT: actually nvm, assuming that the `InternalTopologyBuilder` did not change any actual logic of the original `TopologyBuilder` (it is very hard to tell from the diff file) this is okay to review.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:
```diff
@@ -34,9 +34,9 @@
-import org.apache.kafka.common.metrics.stats.Sum;
 import org.apache.kafka.common.metrics.stats.Max;
 import org.apache.kafka.common.metrics.stats.Rate;
+import org.apache.kafka.common.metrics.stats.Sum;
```
Never know how do you find these: good catch :)

mjsax replied:
Intellij does -- it automatically sort imports if you enable it :)

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/state/Stores.java:
```diff
@@ -36,7 +36,7 @@
  */
 @InterfaceStability.Evolving
-public class Stores {
+public class  Stores {
```
This seems not intentional.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/processor/TopologyBuilder.java:
```diff
@@ -362,9 +155,10 @@ public synchronized final TopologyBuilder addSource(final String name, final Str
-    public synchronized final TopologyBuilder addSource(final AutoOffsetReset offsetReset, final String name, final String... topics) {
-        return addSource(offsetReset, name, null, null, null, topics);
+    public synchronized final TopologyBuilder addSource(final AutoOffsetReset offsetReset,
+                                                        final String name, final String... topics) {
```
nit: if we want to break into multi-lines better do that one parameter per line.

mjsax replied:
Good catch! That was the intent. Just missed this one.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/processor/TopologyBuilder.java:
```diff
@@ -942,26 +645,20 @@ public synchronized final TopologyBuilder addStateStore(final StateStoreSupplier
-            throw new TopologyBuilderException("Source store " + sourceStoreName + " is already added.");
-        }
-        storeToChangelogTopic.put(sourceStoreName, topic);
+    @Deprecated
```
nit: add similar NOTE as in 668 as well, also mentioning it has been deprecated?

guozhangwang replied:
Ditto for all other deprecated functions: mention it should only be used internally and has been deprecated.

guozhangwang reviewed (commented):
Minor comments, otherwise LGTM.

guozhangwang commented:
@mjsax could you rebase?

mjsax commented:
@bbejeck 

>One question (sorry if this is an obvious one): the TopologyBuilder has the InternalTopologyBuilder marked as deprecated. The deprecation indicates using the InternalTopologyBuilder from within TopologyBuilder is deprecated and not the InternalTopologyBuilder itself, correct?

Yes.

mjsax committed:
Github comments

mjsax committed:
Rebased and addressed follow up comments from PR 2301

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6251/
Test PASSed (JDK 7 and Scala 2.11).

mjsax commented:
Did rebase this and addressed "dangling comments" from #2301 
Also converted TopologyDescription and all its nested classed into interfaces (need to update KIP if we keep it this way).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6256/
Test PASSed (JDK 7 and Scala 2.11).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6240/
Test PASSed (JDK 8 and Scala 2.12).

guozhangwang commented:
@mjsax LGTM now. I just left two follow-up comments to your replies on https://github.com/apache/kafka/pull/2301 for 1) package names and 2) printing `none`. LMK.

mjsax committed:
follow ups on commenst from PR 2301

mjsax commented:
Updates with regard to both comments on #2301

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6285/
Test PASSed (JDK 7 and Scala 2.11).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6269/
Test PASSed (JDK 8 and Scala 2.12).

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java:
```diff
@@ -0,0 +1,1491 @@
+        return false;
+    }
+
+    public TopologyDescription describe() {
```
Hmm.. The function is aimed to be a public API in `Topology` in KIP-120 (which I think will be renamed from the current `TopologyBuilder`?), but here it is only used as the function of the `InternalTopologyBuilder` which seems incorrect.

mjsax replied:
> the function is aimed to be a public API in Topology in KIP-120

correct

> which I think will be renamed from the current TopologyBuilder?

No. We need to keep `TopologyBuilder` for backward compatibility (just deprecating it) and add `Topology` next to it. And `Topology` will also have `describe()` in it's public API and than call `InternalTopologyBuilder#describe()` (we might also be able to move this method completely from `InternalTopologyBuilder` to `Topology`, but this would be just an internal refactoring.

guozhangwang replied:
Thanks for the clarification.

Also one thing I notice from the unit tests is that in `StreamThreadTest` we are accessing the builder's `internalTopologyBuilder` object directly which should be deprecated and removed eventually from the `TopologyBuilder`. I believe you are going to update on that with the next follow-up PRs right?

mjsax replied:
I agree, that we will need to do some refactoring there. Not sure though if this should be part of KIP-120 or its own JIRA. (I would prefer own JIRA.)

With `InternalTopologyBuilder` being in package `internals` we can do any kind of refactoring we want without impacting users. There are couple of things we need to think through carefully for this part. I also hope, that we might be able to break some code into more classes that would also simplify testing (hopefully). But this part is hard to design on a whiteboard and must be done in an exploration way "playing" with the code.

guozhangwang replied:
Sounds good.

guozhangwang reviewed (commented):
@mjsax I left a comment on the public APIs of `describe()`, as described in https://cwiki.apache.org/confluence/display/KAFKA/KIP-120%3A+Cleanup+Kafka+Streams+builder+API. Will be merging as is since I understand this is not the last PR of KIP-120 so you may have some ideas about modifying such a bit later, but just as a reminder..

guozhangwang commented:
Merged to trunk.
